File: nhkeasier/edict/deinflect.py
import os.path
from collections import defaultdict
from typing import DefaultDict, Iterator, List, NamedTuple, Set

from .search import Word, search_edict

default_deinflect = os.path.join(os.path.dirname(__file__), 'deinflect.dat')


class Rule(NamedTuple):
    from_: str
    to: str
    type_: int
    reason: str


class Candidate(NamedTuple):
    word: str
    type_: int
    reasons: List[str]


# deinflect.dat countains instructions to remove inflections from words
# the first line is a header
# the next few lines (without '\t') are a string array refereced to later
# the rest are made of four fields separated by '\t'
#   * first field incdicates the suffix to look for in a candidate
#   * second field indicates when the suffix should be remplaced with
#   * third field helps narrowing down the grammatical class of the candidate
#   * fourth field points to the array string and gives a user-friendly
#     explanation of the removed suffix
# type is a bit field where:
#   * bit 0 hints at a 一段 verb ('v1' marker)
#   * bit 1 hints at a 五段 verb (markers starting with 'v5')
#   * bit 2 hints at a い-adjective (marker 'adj-i')
#   * bit 3 hints at a くる verb (marker 'vk')
#   * bit 4 hints at a す or する verb (markers starting with 'vs-')
#   * bit 7 should always be set for words (so that 0xff & wtype != 0 always)
# for a word, type gives a hint of the expected grammatical class of the word
# for a rule, type[0:8] gives the required grammatical class of original word
# for a rule, type[8:16] gives the grammatical class of the resulting word
# thus, the new word has type wtype = rtyle >> 8
class Deinflector:
    """A Deinflector instance applies deinflection rules to normalize a word"""
    def __init__(self, deinflect_data_filename: str = default_deinflect):
        """Populate deinflecting rules from given file"""
        with open(deinflect_data_filename, 'rb') as f:
            lines = iter(f)
            next(lines)  # skip header
            reasons = []  # collect the string array for later resolution
            self.rules: List[Rule] = []
            for byte_line in lines:
                line = byte_line.decode()
                fields = line.strip().split('\t')
                # the header does not indicate the size of the array string; it
                # is simplest to differentiate between the array string and the
                # actual rules by counting the numbers of fields
                if len(fields) == 1:
                    # string array
                    reasons.append(fields[0])
                else:
                    # rule
                    from_, to, stype_, reason = fields
                    type_ = int(stype_)
                    reason = reasons[int(reason)]  # resolve string
                    self.rules.append(Rule(from_, to, type_, reason))

    def __call__(self, word: str) -> List[Candidate]:
        """Iterate through possible deinflections of word (including word)

        Each value is a triplet whose first element is the deinflected word,
        the second element is a mask of possible grammatical classes for the
        word, and the third element is the corresponding reasonning for the
        inflection"""
        candidates = [Candidate(word, 0xff, [])]
        i = 0
        while i < len(candidates):
            candidate = candidates[i]
            for rule in self.rules:
                # check types match
                if candidate.type_ & rule.type_ == 0:
                    continue
                # check suffix matches
                if not candidate.word.endswith(rule.from_):
                    continue
                # append new candidate
                candidates.append(Candidate(
                    word=candidate.word[:-len(rule.from_)] + rule.to,  # replace suffix,
                    type_=rule.type_ >> 8,
                    reasons=candidate.reasons + [rule.reason],
                ))
                # NOTE: could check that new_word is already in candidates
                # Rikaikun merges with previous candidate; if this candidate
                # has already been processed, the new type is ignored
                # Rikaichamp only combines candidates of identical types
            i += 1
        return candidates

    def search_edict(self, fragment: str) -> Iterator[Word]:
        candidates = list(self(fragment))
        subedict: DefaultDict[str, Set[Word]] = defaultdict(set)
        for candidate, _, _ in candidates:
            for word in search_edict(candidate):
                for k in word.readings + word.writings:
                    subedict[k].add(word)
        for candidate, type_, reason in candidates:
            for word in subedict[candidate]:
                if word.get_type() & type_:
                    yield word

File: edict2/src/deinflect.rs
use std::borrow::Cow;
use std::collections::HashMap;

use crate::Error;

#[derive(Clone, Debug, Eq, Hash, PartialEq)]
struct Rule<'a> {
    from: &'a str,
    to: &'a str,
    type_: u32,
    reason: &'a str,
}

#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct Candidate<'a> {
    pub word: Cow<'a, str>,
    pub type_: u32,
}

#[derive(Clone, Debug, Eq, PartialEq)]
struct SuffixToRules<'a>(HashMap<char, (Vec<Rule<'a>>, SuffixToRules<'a>)>);

// deinflect.dat contains instructions to remove inflections from words
// the first line is a header
// the next few lines (without '\t') are a string array referenced to later
// the rest are made of four fields separated by '\t'
//   * first field indicates the suffix to look for in a candidate
//   * second field indicates when the suffix should be replaced with
//   * third field helps narrowing down the grammatical class of the candidate
//   * fourth field points to the array string and gives a user-friendly
//     explanation of the removed suffix
// type is a bit field where:
//   * bit 0 hints at a 一段 verb ('v1' marker)
//   * bit 1 hints at a 五段 verb (markers starting with 'v5')
//   * bit 2 hints at a い-adjective (marker 'adj-i')
//   * bit 3 hints at a くる verb (marker 'vk')
//   * bit 4 hints at a す or する verb (markers starting with 'vs-')
//   * bit 7 should always be set for words (so that 0xff & wtype != 0 always)
// for a word, type gives a hint of the expected grammatical class of the word
// for a rule, type[0:8] gives the required grammatical class of original word
// for a rule, type[8:16] gives the grammatical class of the resulting word
// thus, the new word has type wtype = rtyle >> 8
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Deinflector<'a> {
    suffix_to_rules: SuffixToRules<'a>,
}

impl<'a> Deinflector<'a> {
    pub fn parse(data: &'a str) -> Result<Self, Error> {
        fn aux<'a, I: Iterator<Item = char>>(
            chars: &mut I,
            rule: Rule<'a>,
            cur_suffix_to_rules: &mut SuffixToRules<'a>,
            cur_rules: Option<&mut Vec<Rule<'a>>>,
        ) {
            if let Some(c) = chars.next() {
                if let Some((rules, suffix_to_rules)) = cur_suffix_to_rules.0.get_mut(&c) {
                    aux(chars, rule, suffix_to_rules, Some(rules));
                } else {
                    let mut rules = Vec::new();
                    let mut suffix_to_rules = SuffixToRules(HashMap::new());
                    aux(chars, rule, &mut suffix_to_rules, Some(&mut rules));
                    cur_suffix_to_rules.0.insert(c, (rules, suffix_to_rules));
                }
            } else {
                cur_rules.expect("rules should not be None").push(rule);
            }
        }

        // NOTE: skip(1) for header on first line
        let mut suffix_to_rules = SuffixToRules(HashMap::new());
        let mut reasons = Vec::new();
        for (lineno, line) in data.lines().enumerate().skip(1) {
            let lineno = lineno + 1;
            let fields: Vec<&str> = line.split('\t').collect();
            match fields[..] {
                [_] => reasons.push(line),
                [from, to, type_, reason] => {
                    let type_: u32 = type_.parse().map_err(|_| Error::ParseInteger { lineno })?;
                    let reason: usize =
                        reason.parse().map_err(|_| Error::ParseInteger { lineno })?;
                    let reason = reasons[reason];
                    let rule = Rule {
                        from,
                        to,
                        type_,
                        reason,
                    };
                    aux(&mut from.chars().rev(), rule, &mut suffix_to_rules, None)
                }
                _ => panic!("unexpected line {line}"),
            }
        }
        Ok(Deinflector { suffix_to_rules })
    }

    pub fn deinflect<'b, 'c>(
        &self,
        word: &'c str,
        vec: &'b mut Vec<Candidate<'c>>,
    ) -> Iter<'_, 'b, 'c> {
        vec.clear();
        vec.push(Candidate {
            word: Cow::Borrowed(word),
            type_: 0xff,
        });
        Iter {
            deinflector: self,
            candidates: vec,
        }
    }
}

pub struct Iter<'a, 'b, 'c> {
    deinflector: &'a Deinflector<'a>,
    candidates: &'b mut Vec<Candidate<'c>>,
}

impl<'a, 'b, 'c: 'a> Iterator for Iter<'a, 'b, 'c> {
    type Item = Candidate<'a>;
    fn next(&mut self) -> Option<Self::Item> {
        let candidate = self.candidates.pop()?;
        let mut cur_suffix_to_rules = &self.deinflector.suffix_to_rules;
        for c in candidate.word.chars().rev() {
            if let Some((rules, suffix_to_rules)) = cur_suffix_to_rules.0.get(&c) {
                cur_suffix_to_rules = suffix_to_rules;
                for rule in rules {
                    if candidate.type_ & rule.type_ == 0 {
                        continue;
                    }
                    let prefix_len = candidate.word.bytes().len() - rule.from.bytes().len();
                    let prefix = &candidate.word[..prefix_len];
                    let mut word = String::with_capacity(prefix_len + rule.to.as_bytes().len());
                    word.push_str(prefix);
                    word.push_str(rule.to);
                    self.candidates.push(Candidate {
                        word: Cow::Owned(word),
                        type_: rule.type_ >> 8,
                    })
                }
            } else {
                break;
            }
        }
        Some(candidate)
    }
}
